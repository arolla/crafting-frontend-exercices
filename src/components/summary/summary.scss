@use '@shared/styles/elements';
@use '@shared/styles/wrappers';
@use "@shared/styles/variables";

.summary {
  @extend %main-wrapper;
  @extend %separator;
}

.summary__title {
  @extend %medium-title;
}

.summary__info {
  @extend %content-wrapper;

  p {
    font-size: variables.$medium-text;
  }
}
