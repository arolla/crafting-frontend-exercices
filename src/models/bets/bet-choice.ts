export type BetChoice = 'TEAM_1_WINS' | 'DRAW' | 'TEAM_2_WINS'
