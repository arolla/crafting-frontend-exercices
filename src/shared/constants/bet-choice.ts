import type { BetChoice } from '@models/bets/bet-choice'

export const TEAM_1_WINS: BetChoice = 'TEAM_1_WINS'
export const DRAW: BetChoice = 'DRAW'
export const TEAM_2_WINS: BetChoice = 'TEAM_2_WINS'
