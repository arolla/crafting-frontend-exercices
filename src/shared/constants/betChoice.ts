import { BetChoice } from '../../models'

export const TEAM_1_WINS: BetChoice = 'TEAM_1_WINS'
export const DRAW: BetChoice = 'DRAW'
export const TEAM_2_WINS: BetChoice = 'TEAM_2_WINS'
