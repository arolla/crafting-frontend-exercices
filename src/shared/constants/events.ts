export const UPDATE_STAKE = 'UPDATE_STAKE'
