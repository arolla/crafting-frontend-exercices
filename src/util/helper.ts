export const withDelay = <T>(
  fn: () => Promise<T>,
  delay?: number,
): Promise<T> =>
  new Promise((resolve, reject) => {
    setTimeout(async () => {
      fn().then(resolve, reject)
    }, delay)
  })
