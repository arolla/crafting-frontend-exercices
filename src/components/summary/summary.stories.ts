import type { Meta, StoryObj } from '@storybook/web-components'
import { Summary } from './summary'

Summary.register()

const meta: Meta<Summary> = {
  title: 'Components/Summary',
  component: 'arl-summary',
  argTypes: {
    stake: {
      control: 'number',
    },
  },
  args: {
    stake: 100,
    betsSlip: [
      {
        gameId: 'gameId',
        selectedChoice: 'TEAM_1_WINS',
        selectedOdds: 2.98,
      },
      {
        gameId: 'gameId',
        selectedChoice: 'DRAW',
        selectedOdds: 1.64,
      },
    ],
  },
}

export default meta

type Story = StoryObj<Summary>

export const Default: Story = {}
