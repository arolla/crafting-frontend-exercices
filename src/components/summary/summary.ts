import { computePotentialGains } from '@business/business.rules'
import type { BetSlip } from '@models/bets/bet-slip'
import { staticImplements } from '@util/decorators.helper.'
import { parseNumber } from '@util/string.helper'
import { WebComponent, type WebComponentConstructor } from '@util/web-component'
import { html } from 'common-tags'
import css from './summary.scss?inline'

@staticImplements<WebComponentConstructor>()
export class Summary extends WebComponent {
  static register() {
    WebComponent.register('arl-summary', Summary)
  }

  #betsSlip: BetSlip[] = []
  #stake = 0

  get stake() {
    return this.#stake
  }

  get betsSlip() {
    return this.#betsSlip
  }

  set betsSlip(betsSlip: BetSlip[]) {
    this.#betsSlip = betsSlip
    this.render()
  }

  set stake(stake: number) {
    if (this.#stake === stake) {
      return
    }
    this.#stake = stake
    this.render()
  }

  buildTemplate() {
    if (!this.betsSlip.length) {
      return ''
    }
    const potentialGains = computePotentialGains(this.stake, this.betsSlip)
    return html`
      <style>${css}</style>
      <div class="summary">
        <h3 class="summary__title">Betting summary</h3>
        <div class="summary__info">
          <p class="summary__info--bets-slip">Number of bets placed: ${this.betsSlip.length}</p>
          <p class="summary__info--potential-gains">Potential gain: ${potentialGains} â‚¬</p>
        </div>
      </div>
    `
  }

  static get observedAttributes() {
    return ['stake']
  }

  attributeChangedCallback(name: string, _oldValue: string, newValue: string) {
    switch (name) {
      case 'stake': {
        const stake = parseNumber(newValue)
        if (this.stake !== stake) {
          this.#stake = stake
          this.render()
        }
        break
      }
      default:
    }
  }
}
