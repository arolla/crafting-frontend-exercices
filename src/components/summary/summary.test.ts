import { beforeAll, describe, expect, it } from '@jest/globals'
import type { BetSlip } from '@models/bets/bet-slip'
import { DRAW, TEAM_1_WINS } from '@shared/constants/bet-choice'
import { stringify } from '@util/string.helper'
import { Summary } from './summary'

const DUMMY_BETS_SLIP: BetSlip[] = [
  {
    gameId: 'id1',
    selectedChoice: TEAM_1_WINS,
    selectedOdds: 1.52,
  },
  {
    gameId: 'id2',
    selectedChoice: DRAW,
    selectedOdds: 1.12,
  },
]
const DUMMY_STAKE = 100

describe('Bets-Summary Component', () => {
  beforeAll(() => {
    Summary.register()
  })
  test('should render correctly summary when user made bets slip with 100â‚¬ stake', () => {
    // Given
    const summary = new Summary()
    summary.render()
    // When
    summary.betsSlip = DUMMY_BETS_SLIP
    summary.setAttribute('stake', `${DUMMY_STAKE}`)
    // Then
    expect(summary?.shadowRoot?.querySelector('.summary')).toMatchSnapshot()
  })
})
