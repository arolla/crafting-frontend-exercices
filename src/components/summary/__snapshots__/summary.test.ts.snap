// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bets-Summary Component should render correctly summary when user made bets slip with 100€ stake 1`] = `
<div
  class="summary"
>
  
  
  <h3
    class="summary__title"
  >
    Betting summary
  </h3>
  
  
  <div
    class="summary__info"
  >
    
    
    <p
      class="summary__info--bets-slip"
    >
      Number of bets placed: 2
    </p>
    
    
    <p
      class="summary__info--potential-gains"
    >
      Potential gain:  €
    </p>
    
  
  </div>
  

</div>
`;
